File: ReturnTracker/Features/EmailSync/GmailAPIClient.swift
import Foundation

struct GmailAPIClient {
    func fetchRecentPurchaseEmails(accessToken: String) async throws -> [EmailMessageSnapshot] {
        var components = URLComponents(string: "https://gmail.googleapis.com/gmail/v1/users/me/messages")!
        components.queryItems = [
            URLQueryItem(name: "maxResults", value: "30"),
            URLQueryItem(
                name: "q",
                value: "newer_than:90d {subject:주문 subject:결제 subject:구매 subject:order subject:receipt} -subject:취소"
            )
        ]
        let list: GmailMessageList = try await request(components.url!, accessToken: accessToken)

        var snapshots: [EmailMessageSnapshot] = []
        for message in list.messages ?? [] {
            let url = URL(
                string: "https://gmail.googleapis.com/gmail/v1/users/me/messages/\(message.id)?format=full"
            )!
            let detail: GmailMessage = try await request(url, accessToken: accessToken)
            if let snapshot = detail.snapshot {
                snapshots.append(snapshot)
            }
        }
        return snapshots
    }

    private func request<Response: Decodable>(
        _ url: URL,
        accessToken: String
    ) async throws -> Response {
        var request = URLRequest(url: url)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            let status = (response as? HTTPURLResponse)?.statusCode ?? -1
            let message = String(data: data, encoding: .utf8)?.prefix(160) ?? ""
            throw GmailAPIError.requestFailed(status: status, message: String(message))
        }
        return try JSONDecoder().decode(Response.self, from: data)
    }
}

enum GmailAPIError: LocalizedError {
    case requestFailed(status: Int, message: String)

    var errorDescription: String? {
        switch self {
        case let .requestFailed(status, _):
            status == 401
                ? "Gmail 권한이 만료되었습니다. 다시 동기화해 주세요."
                : "Gmail 요청에 실패했습니다. (\(status))"
        }
    }
}

private struct GmailMessageList: Decodable {
    let messages: [GmailMessageReference]?
}

private struct GmailMessageReference: Decodable {
    let id: String
}

private struct GmailMessage: Decodable {
    let id: String
    let internalDate: String?
    let payload: GmailPayload?

    var snapshot: EmailMessageSnapshot? {
        guard let payload else { return nil }
        var headers: [String: String] = [:]
        for header in payload.headers ?? [] where headers[header.name.lowercased()] == nil {
            headers[header.name.lowercased()] = header.value
        }
        var plainBodies: [String] = []
        var htmlBodies: [String] = []
        payload.collectBodies(plain: &plainBodies, html: &htmlBodies)
        let body = plainBodies.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
        let resolvedBody = body.isEmpty
            ? htmlBodies.joined(separator: "\n").strippingHTML
            : body

        return EmailMessageSnapshot(
            id: id,
            subject: headers["subject"] ?? "",
            from: headers["from"] ?? "",
            receivedAt: internalDate.flatMap(Double.init).map { Date(timeIntervalSince1970: $0 / 1000) } ?? .now,
            body: resolvedBody
        )
    }
}

private struct GmailPayload: Decodable {
    let mimeType: String?
    let headers: [GmailHeader]?
    let body: GmailBody?
    let parts: [GmailPayload]?

    func collectBodies(plain: inout [String], html: inout [String]) {
        if let encoded = body?.data,
           let decoded = encoded.base64URLDecodedString,
           !decoded.isEmpty {
            if mimeType?.hasPrefix("text/plain") == true {
                plain.append(decoded)
            } else if mimeType?.hasPrefix("text/html") == true {
                html.append(decoded)
            }
        }
        for part in parts ?? [] {
            part.collectBodies(plain: &plain, html: &html)
        }
    }
}

private struct GmailHeader: Decodable {
    let name: String
    let value: String
}

private struct GmailBody: Decodable {
    let data: String?
}

private extension String {
    var base64URLDecodedString: String? {
        var value = replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        value += String(repeating: "=", count: (4 - value.count % 4) % 4)
        guard let data = Data(base64Encoded: value) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    var strippingHTML: String {
        replacingMatches("<style[^>]*>.*?</style>", with: " ", dotMatchesLineSeparators: true)
            .replacingMatches("<script[^>]*>.*?</script>", with: " ", dotMatchesLineSeparators: true)
            .replacingMatches("<br\\s*/?>|</p>|</div>|</tr>", with: "\n")
            .replacingMatches("<[^>]+>", with: " ")
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingMatches("[ \\t]+", with: " ")
            .replacingMatches("\\n{3,}", with: "\n\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    func replacingMatches(
        _ pattern: String,
        with replacement: String,
        dotMatchesLineSeparators: Bool = false
    ) -> String {
        var options: NSRegularExpression.Options = .caseInsensitive
        if dotMatchesLineSeparators {
            options.insert(.dotMatchesLineSeparators)
        }
        guard let expression = try? NSRegularExpression(pattern: pattern, options: options) else {
            return self
        }
        return expression.stringByReplacingMatches(
            in: self,
            range: NSRange(startIndex..., in: self),
            withTemplate: replacement
        )
    }
}

File: android/app/src/main/java/com/returntracker/android/email/GmailApiClient.kt
package com.returntracker.android.email

import android.util.Base64
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.time.Instant
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject

class GmailApiClient {
    suspend fun fetchRecentPurchaseEmails(accessToken: String): List<EmailMessageSnapshot> =
        withContext(Dispatchers.IO) {
            val query = "newer_than:90d {subject:주문 subject:결제 subject:구매 subject:order subject:receipt} -subject:취소"
            val encodedQuery = URLEncoder.encode(query, StandardCharsets.UTF_8.name())
            val listJson = getJson(
                url = "$BASE_URL/users/me/messages?maxResults=$MAX_RESULTS&q=$encodedQuery",
                accessToken = accessToken,
            )
            val messages = listJson.optJSONArray("messages") ?: return@withContext emptyList()

            buildList {
                for (index in 0 until messages.length()) {
                    val id = messages.optJSONObject(index)?.optString("id").orEmpty()
                    if (id.isBlank()) continue
                    val json = getJson(
                        url = "$BASE_URL/users/me/messages/$id?format=full",
                        accessToken = accessToken,
                    )
                    parseMessage(json)?.let(::add)
                }
            }
        }

    private fun parseMessage(json: JSONObject): EmailMessageSnapshot? {
        val id = json.optString("id")
        if (id.isBlank()) return null

        val payload = json.optJSONObject("payload") ?: return null
        val headers = buildMap {
            val array = payload.optJSONArray("headers") ?: return@buildMap
            for (index in 0 until array.length()) {
                val header = array.optJSONObject(index) ?: continue
                put(header.optString("name").lowercase(), header.optString("value"))
            }
        }
        val plainBodies = mutableListOf<String>()
        val htmlBodies = mutableListOf<String>()
        collectBodies(payload, plainBodies, htmlBodies)
        val body = plainBodies.joinToString("\n").ifBlank {
            htmlBodies.joinToString("\n").stripHtml()
        }

        return EmailMessageSnapshot(
            id = id,
            subject = headers["subject"].orEmpty(),
            from = headers["from"].orEmpty(),
            receivedAt = json.optString("internalDate").toLongOrNull()?.let(Instant::ofEpochMilli)
                ?: Instant.now(),
            body = body,
        )
    }

    private fun collectBodies(
        part: JSONObject,
        plainBodies: MutableList<String>,
        htmlBodies: MutableList<String>,
    ) {
        val mimeType = part.optString("mimeType")
        val encoded = part.optJSONObject("body")?.optString("data").orEmpty()
        if (encoded.isNotBlank()) {
            val decoded = runCatching {
                val bytes = Base64.decode(encoded, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
                String(bytes, Charsets.UTF_8)
            }.getOrNull()
            if (!decoded.isNullOrBlank()) {
                when {
                    mimeType.startsWith("text/plain") -> plainBodies += decoded
                    mimeType.startsWith("text/html") -> htmlBodies += decoded
                }
            }
        }

        val parts = part.optJSONArray("parts") ?: return
        for (index in 0 until parts.length()) {
            parts.optJSONObject(index)?.let { collectBodies(it, plainBodies, htmlBodies) }
        }
    }

    private fun getJson(url: String, accessToken: String): JSONObject {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 20_000
            setRequestProperty("Authorization", "Bearer $accessToken")
            setRequestProperty("Accept", "application/json")
        }

        return try {
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val response = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
            if (status !in 200..299) {
                throw IOException("Gmail API 요청 실패 ($status): ${response.take(160)}")
            }
            JSONObject(response)
        } finally {
            connection.disconnect()
        }
    }

    private fun String.stripHtml(): String =
        replace(Regex("<style[^>]*>.*?</style>", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL)), " ")
            .replace(Regex("<script[^>]*>.*?</script>", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL)), " ")
            .replace(Regex("<br\\s*/?>|</p>|</div>|</tr>", RegexOption.IGNORE_CASE), "\n")
            .replace(Regex("<[^>]+>"), " ")
            .replace("&nbsp;", " ")
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace(Regex("[ \\t]+"), " ")
            .replace(Regex("\\n{3,}"), "\n\n")
            .trim()

    private companion object {
        const val BASE_URL = "https://gmail.googleapis.com/gmail/v1"
        const val MAX_RESULTS = 30
    }
}
